nixos/modules/config/ldso.nix

```diff
@@ -0,0 +1,61 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) last splitString mkOption types mdDoc mkMerge escapeShellArg mkIf;
+
+  libDir = pkgs.stdenv.hostPlatform.libDir;
+  ldsoBasename = last (splitString "/" pkgs.stdenv.cc.bintools.dynamicLinker);
+
+  pkgs32 = pkgs.pkgsi686Linux;
+  libDir32 = pkgs32.stdenv.hostPlatform.libDir;
+  ldsoBasename32 = last (splitString "/" pkgs32.stdenv.cc.bintools.dynamicLinker);
+in {
+  options = {
+    environment.ldso = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = mdDoc ''
+        The executable to link into the normal FHS location of the ELF loader.
+
+        This is set by default by `environment.stub-ld.enable`. Disable `stub-ld` if you wish to set this.
+      '';
+    };
+
+    environment.ldso32 = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = mdDoc ''
+        The executable to link into the normal FHS location of the 32-bit ELF loader.
+
+        This currently only works on x86_64 architectures.
+
+        This is set by default by `environment.stub-ld.enable`. Disable `stub-ld` if you wish to set this.
+      '';
+    };
+  };
+
+  config = {
+    assertions = [
+      { assertion = isNull config.environment.ldso32 || pkgs.stdenv.isx86_64;
+        message = "Option environment.ldso32 currently only works on x86_64.";
+      }
+    ];
+
+    systemd.tmpfiles.rules = mkMerge [
+      (if isNull config.environment.ldso then [
+        "r /${libDir}/${ldsoBasename} - - - - -"
+      ] else [
+        "d /${libDir} 0755 root root - -"
+        "L+ /${libDir}/${ldsoBasename} - - - - ${config.environment.ldso}"
+      ])
+      (mkIf pkgs.stdenv.isx86_64 (if isNull config.environment.ldso32 then [
+        "r /${libDir32}/${ldsoBasename32} - - - - -"
+      ] else [
+        "d /${libDir32} 0755 root root - -"
+        "L+ /${libDir32}/${ldsoBasename32} - - - - ${config.environment.ldso32}"
+      ]))
+    ];
+  };
```

Comment:
This is hard to read. Is there a reason why this isn't a ternary with some `optionals`?

Reply:
I could use `optionals` instead of `mkMerge` and `mkIf`, I just didn't think to when switching from activation scripts.

I'm not sure what you mean by "a ternary and some optionals", though. Seems like there need to be 2 ternaries. Even if you could technically avoid them by replicating conditions across optionals, that's just not good code imo. I also want the similarity between the main arch and the 32bit code to be obvious, so using a ternary for one and not the other isn't good.

Reply:
Better?

Reply:
I had missed some of the complexity in the multiarch part.

What would simplify this a lot would be an enable option to get rid of the isNulls; enabling both archs.

Reply:
I'd rather make it easy for the user at the expense of the module's readability than the other way around. This exact model of a nullable path option is already used for, e.g. `/usr/bin/env`, and seems intuitive enough.